### ISSUE TITLE
Some questions about the code

### ISSUE THREAD

**Issue by shallweiwei** [fenglinglwb/MAT#65]
Hi! @fenglinglwb 
I'm trying to understand your code line by line and I've faced some difficulties.```
In the class Conv2dLayerPartial ,from the name it looks like partial convolution,but it seems it's not a partial convolution.
https://github.com/fenglinglwb/MAT/blob/d273d891ecdad2e1df106516423a75bc45b2d800/networks/mat.py#L99
You can see that there is number 9 which devided by "update_mask" ,I don't understand what this whole class is doing.After some loops here ,it always gives a matrix which all elments are one.And so this line of code doesn't make effects.
https://github.com/fenglinglwb/MAT/blob/d273d891ecdad2e1df106516423a75bc45b2d800/networks/mat.py#L161
So I am confused about the mask updating strategy.
Thanks for your helping.

**Comment by fenglinglwb** [comment]
As for the partial convolution, we refer to
https://github.com/NVIDIA/partialconv/blob/610d373f35257887d45adae84c86d0ce7ad808ec/models/partialconv2d.py#L64

**Comment by shallweiwei** [comment]
Thanks you.I understand partial convolution code now.
And I'll try to unterstand the rest of the code

**Comment by shallweiwei** [comment]
Hi  @fenglinglwb 
Maybe my understanding of shift window is poor.I'm still confused about the mask updating code
![2023-02-06 17-30-24屏幕截图](https://user-images.githubusercontent.com/84014580/216936277-712f3af5-bdb8-49c3-8134-bb682b1128e1.png)
In your code the mask window shifts first
https://github.com/fenglinglwb/MAT/blob/d273d891ecdad2e1df106516423a75bc45b2d800/networks/mat.py#L264
And then is window partition,next when we calculate attention.In class WindowAttention ,I believe this is the code of mask updating.
https://github.com/fenglinglwb/MAT/blob/d273d891ecdad2e1df106516423a75bc45b2d800/networks/mat.py#L164

I think as the window is shifted ,mask should not update here.Maybe it should update after the shift is reverse

**Comment by shallweiwei** [comment]
Is there anyone who can help me?

**Comment by fenglinglwb** [comment]
The mask and feature should be shifted simultaneously. We must shift the mask, update the mask, and finally re-shift the mask back to its original state.

**Comment by shallweiwei** [comment]
@fenglinglwb But after the mask is shifted,the mask is in the wrong window.This is a way for calculating window_attention easily.
![image](https://user-images.githubusercontent.com/84014580/218072294-397f7c94-1a05-4a91-8da6-5f31a8fec431.png)

**Comment by shallweiwei** [comment]
这里我使用中文，左上角的到了右下角，但是计算attention的时候swin transformer是专门为了featrue做了一个attention_mask，但你这个普通的mask可能就不对了

**Comment by fenglinglwb** [comment]
Based on the Swin, we just add an additional mask. As shown below, we have mask_windows and mask two variables.
https://github.com/fenglinglwb/MAT/blob/d273d891ecdad2e1df106516423a75bc45b2d800/networks/mat.py#L140

**Comment by shallweiwei** [comment]
是的，我理解这个，但是这个mask_windows,因为你已经shift过了，这样的话如果左上角的mask有效，这部分会被移到右下方，这样的话按照这个更新的代码，你的右下角窗口反而会被更新，和你的本意是不一样的。

**Comment by fenglinglwb** [comment]
Sure. For most windows, we update the pixels based on our strategy. But for the corner ones, we just follow Swin which also involves non-adjacent pixels for convenience. You may improve our strategy following your ideas.

**Comment by shallweiwei** [comment]
非常感谢！
